integration/fuse/src/main/java/alluxio/fuse/file/FuseFileInOrOutStream.java

```diff
@@ -0,0 +1,165 @@
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An implementation for {@link FuseFileStream} for read only or write only workloads.
+ * Fuse can open file for reading and writing concurrently but Alluxio only support
+ * read-only or write-only workloads. This class will be used as write only stream
+ * if O_TRUNC flag is provided or first operation is write() or truncate(),
+ * will be used as read only stream otherwise.
+ */
+@ThreadSafe
+public class FuseFileInOrOutStream implements FuseFileStream {
+  private static final Logger LOG = LoggerFactory.getLogger(FuseFileInOrOutStream.class);
+  private final FileSystem mFileSystem;
+  private final AuthPolicy mAuthPolicy;
+  private final AlluxioURI mUri;
+  private final long mMode;
+  private final Optional<URIStatus> mStatus;
+
+  // underlying reed-only or write-only stream
+  private FuseFileStream mStream;
+
+  /**
+   * Creates a {@link FuseFileInOrOutStream}.
+   *
+   * @param fileSystem the Alluxio file system
+   * @param authPolicy the Authentication policy
+   * @param uri the alluxio uri
+   * @param flags the fuse create/open flags
+   * @param mode the filesystem mode, -1 if not set
+   * @param status the uri status, null if not uri does not exist
+   * @return a {@link FuseFileInOrOutStream}
+   */
+  public static FuseFileInOrOutStream create(FileSystem fileSystem, AuthPolicy authPolicy,
+      AlluxioURI uri, int flags, long mode, Optional<URIStatus> status)
+      throws IOException, ExecutionException, AlluxioException {
+    Preconditions.checkNotNull(fileSystem);
+    Preconditions.checkNotNull(uri);
+    Preconditions.checkNotNull(status);
+    boolean truncate = AlluxioFuseOpenUtils.containsTruncate(flags);
+    if (status.isPresent() && truncate) {
+      fileSystem.delete(uri);
+      status = Optional.empty();
+      LOG.debug(String.format("Open path %s with flag 0x%x for overwriting. "
+          + "Alluxio deleted the old file and created a new file for writing", uri, flags));
+    }
+    if (!status.isPresent()) {
+      FuseFileOutStream stream = FuseFileOutStream.create(fileSystem, authPolicy,
+          uri, OpenFlags.O_WRONLY.intValue(), mode, status);
+      return new FuseFileInOrOutStream(fileSystem, authPolicy, stream, status, uri, mode);
+    }
+    // Left for next operation to decide read-only or write-only mode
+    // read-only: open(READ_WRITE) existing file - read()
+    // write-only: open(READ_WRITE) existing file - truncate(0) - write()
+    return new FuseFileInOrOutStream(fileSystem, authPolicy, null, status, uri, mode);
+  }
+
+  private FuseFileInOrOutStream(FileSystem fileSystem, AuthPolicy authPolicy,
+      @Nullable FuseFileStream stream, Optional<URIStatus> status, AlluxioURI uri, long mode) {
+    mFileSystem = fileSystem;
+    mAuthPolicy = authPolicy;
+    mStream = stream;
+    mStatus = status;
+    mUri = uri;
+    mMode = mode;
+  }
+
+  @Override
+  public synchronized int read(ByteBuffer buf, long size, long offset)
```

Comment:
Using `synchronized` on all method calls is very inefficient. It's probably better to address the underlying problem. Please create an issue as a follow up.